File: tests/src/com.typesafe.akka/akka-protobuf-v3_3/2.9.0-M2/src/test/scala/com_typesafe_akka/akka_protobuf_v3_3/ByteBufferWriterTest.scala
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package com_typesafe_akka.akka_protobuf_v3_3

import java.io.ByteArrayOutputStream
import java.lang.Void.TYPE
import java.lang.invoke.MethodHandle
import java.lang.invoke.MethodHandles
import java.lang.invoke.MethodType.methodType
import java.nio.ByteBuffer

import akka.protobufv3.internal.ByteString
import akka.protobufv3.internal.UnsafeByteOperations
import org.junit.jupiter.api.Assertions.assertArrayEquals
import org.junit.jupiter.api.Test

class ByteBufferWriterTest {
  @Test
  def writesDirectByteBufferRangeToOutputStream(): Unit = {
    val sourceBytes: Array[Byte] = Array[Byte](10, 11, 12, 13, 14, 15)
    val directBuffer: ByteBuffer = ByteBuffer.allocateDirect(sourceBytes.length)
    directBuffer.put(sourceBytes)
    directBuffer.flip()

    val byteString: ByteString = UnsafeByteOperations.unsafeWrap(directBuffer)
    val output: ByteArrayOutputStream = new ByteArrayOutputStream()

    byteStringRangeWriter.invokeWithArguments(
      byteString,
      output,
      java.lang.Integer.valueOf(1),
      java.lang.Integer.valueOf(4)
    )

    assertArrayEquals(Array[Byte](11, 12, 13, 14), output.toByteArray)
  }

  private def byteStringRangeWriter: MethodHandle = {
    val lookup: MethodHandles.Lookup = MethodHandles.privateLookupIn(
      classOf[ByteString],
      MethodHandles.lookup()
    )
    lookup.findVirtual(
      classOf[ByteString],
      "writeTo",
      methodType(
        TYPE,
        classOf[java.io.OutputStream],
        java.lang.Integer.TYPE,
        java.lang.Integer.TYPE
      )
    )
  }
}

File: tests/src/com.google.protobuf/protobuf-javalite/4.0.0-rc-1/src/test/java/com_google_protobuf/protobuf_javalite/ByteBufferWriterTest.java
/*
 * Copyright and related rights waived via CC0
 *
 * You should have received a copy of the CC0 legalcode along with this
 * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
package com_google_protobuf.protobuf_javalite;

import static org.assertj.core.api.Assertions.assertThat;

import com.google.protobuf.ByteString;
import java.io.ByteArrayOutputStream;
import java.io.OutputStream;
import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import java.lang.invoke.MethodType;
import java.nio.ByteBuffer;
import org.junit.jupiter.api.Test;

public class ByteBufferWriterTest {
    @Test
    void writesDirectByteBufferToOutputStreamWithoutChangingPosition() throws Exception {
        byte[] payload = new byte[] {0x10, 0x20, 0x30, 0x40};
        ByteBuffer source = ByteBuffer.allocateDirect(payload.length + 2);
        source.put((byte) 0x00);
        source.put(payload);
        source.put((byte) 0x7f);
        source.flip();
        source.position(1);
        source.limit(1 + payload.length);
        ByteArrayOutputStream output = new ByteArrayOutputStream();

        writeDirectBuffer(source, output);

        assertThat(output.toByteArray()).containsExactly(payload);
        assertThat(source.position()).isEqualTo(1);
    }

    private static void writeDirectBuffer(ByteBuffer source, OutputStream output) throws Exception {
        Class<?> writerType = Class.forName(
                "com.google.protobuf.ByteBufferWriter",
                true,
                ByteString.class.getClassLoader());
        MethodHandle write = MethodHandles.privateLookupIn(writerType, MethodHandles.lookup())
                .findStatic(
                        writerType,
                        "write",
                        MethodType.methodType(void.class, ByteBuffer.class, OutputStream.class));
        try {
            write.invoke(source, output);
        } catch (Error | RuntimeException e) {
            throw e;
        } catch (Throwable e) {
            throw new IllegalStateException("ByteBufferWriter.write failed", e);
        }
    }
}
